plugins/postcss-color-function/src/index.ts

```diff
@@ -0,0 +1,122 @@
+import { hasSupportsAtRuleAncestor } from './has-supports-at-rule-ancestor';
+import valueParser from 'postcss-value-parser';
+import type { ParsedValue, FunctionNode } from 'postcss-value-parser';
+import type { Declaration, Result } from 'postcss';
+import { onCSSFunctionSRgb } from './on-css-function';
+import { hasFallback } from './has-fallback-decl';
+import type { PluginCreator } from 'postcss';
+
+/** Transform color() function in CSS. */
+const postcssPlugin: PluginCreator<{ preserve: boolean }> = (opts?: { preserve: boolean }) => {
+	const preserve = 'preserve' in Object(opts) ? Boolean(opts.preserve) : false;
+
+	return {
+		postcssPlugin: 'postcss-color-function',
+		Declaration: (decl: Declaration, { result }: { result: Result }) => {
+			if (hasFallback(decl)) {
+				return;
+			}
+
+			if (hasSupportsAtRuleAncestor(decl)) {
+				return;
+			}
+
+			const originalValue = decl.value;
+			if (originalValue.indexOf('color(') === -1) {
+				return;
+			}
+
+			const modified = modifiedValues(originalValue, decl, result);
+			if (typeof modified === 'undefined') {
+				return;
+			}
+
+			if (preserve) {
+				decl.cloneBefore({ value: modified });
+			} else {
+				decl.value = modified;
+			}
+		},
+		RuleExit: (rule, { postcss }) => {
+			if (!preserve) {
+				return;
+			}
+
+			const atSupportsRules = [];
+			const variableNames = new Set<string>();
+
+			rule.each((decl) => {
+				if (decl.type !== 'decl') {
+					return;
+				}
+
+				if (!decl.variable) {
+					return;
+				}
+
+				if (!variableNames.has(decl.prop.toString())) {
+					variableNames.add(decl.prop.toString());
+					return;
+				}
+
+				const atSupports = postcss.atRule({ name: 'supports', params: `(${decl.prop}: ${decl.value})`, source: rule.source });
+				const parentClone = rule.clone();
+				parentClone.removeAll();
+
+				parentClone.append(decl.clone());
+				decl.remove();
+
+				atSupports.append(parentClone);
+				atSupportsRules.push(atSupports);
+			});
+
+			if (atSupportsRules.length === 0) {
+				return;
+			}
+
+			atSupportsRules.reverse().forEach((atSupports) => {
+				rule.after(atSupports);
+			});
+		},
```

Comment:
I am unsure about this bit.

```css
:root {
  --one-var: red;
  --one-var: color(srgb 0.64331 var(--point-5) 0.16771 / var(--opacity-50));
}
```

As setting the same custom property twice doesn't make sense the second is wrapped in `@supports`.

It is assumed that one is a fallback created by a plugin with `preserve: true`;

It is however not limited to `color`.

Should this be a separate plugin?
Or is it ok if we have this logic in multiple plugins?

I don't see how it can cause harm, I am just not a 100% happy with it.

Reply:
This is an interesting thought and I feel like you. How does it sound about doing something such as what you did to avoid infinite loops? Like an interim step that another plugin can then take over (only within preset env)

Reply:
From a technical standpoint an interim step would not be needed.

Maybe one of these :
- separate plugin and list it as a peer dependency in package.json?
- separate plugin and make each that needs it a tiny plugin pack with the option to disable

Reply:
A separate plugin would remove the duplication (each value transform plugin needs this logic).

And it would also resolve all the bug reports we have seen in preset-env and plugins about broken vars.

I think it is ok to have "utility" plugins in preset-env that cleanup and fix css.
As long as these are limited to such technical issues caused by plugin transforms and do not aim to fix stylesheet author issues.